concurrency/NonMemberPreferencesMigrationHelper.java

```diff
@@ -0,0 +1,397 @@
+ * there is a logged in user, by synchronizing settings between external
+ * storage directory and SharedPreferences. The former stores non-member settings
+ * to be migrated on top of SharedPreferences once there is a logged in user.
+ * The settings stored in SharedPreferences are used as the source of truth
+ * for as long as the device is in a non-member mode. In short, this class
+ * maintains the following two way synchronization:
+ *     When in non-member mode:
+ *         continuous: SharedPreferences -> External Storage Directory
+ *     When in member mode:
+ *         one-time: External Storage Directory -> SharedPreferences
+ *
+ * All settings data that is stored in external storage directory is serialized
+ * as JSON. The corresponding backing file is shared across all settings, which
+ * is why it is imperative that each has a unique key.
+ */
+public class NonMemberPreferencesMigrationHelper {
+
+    private static final String TAG = NonMemberPreferencesMigrationHelper.class.getSimpleName();
+    private static final String ABS_PATH_FORMAT = "%s/com.amazon.photos.NonMemberPreferencesMigrationHelper.json";
+
+    @NonNull private final ObjectMapper mapper;
+    @NonNull private final File file;
+    @NonNull private final TypeReference<ConcurrentHashMap<String, Object>> typeRef;
+    @NonNull private final CountDownLatch loadedFromDisk;
+    @NonNull private ConcurrentHashMap<String, Object> map;
+    @NonNull NonMemberSettingsMigrationMetricsHelper metricsHelper;
+
+    private Disposable commitToDiskDisposable;
+    private String accountIdSettingsSynchronizationDoneFor;
+
+    public NonMemberPreferencesMigrationHelper(@Nullable final Profiler profiler) {
+        mapper = new ObjectMapper();
+        typeRef = new TypeReference<ConcurrentHashMap<String, Object>>() { };
+        file = new File(String.format(ABS_PATH_FORMAT,
+                Environment.getExternalStorageDirectory().getAbsolutePath()));
+        loadedFromDisk = new CountDownLatch(1);
+        map = new ConcurrentHashMap<>();
+        metricsHelper = new NonMemberSettingsMigrationMetricsHelper(profiler);
+
+        loadFromDisk();
+    }
+
+    @VisibleForTesting
+    public NonMemberPreferencesMigrationHelper() {
+        this(null);
+    }
+
+    /**
+     * Retrieve a String value from the persistent preferences.
+     * @param key The name of the preference to retrieve.
+     * @param defaultValue Value to return if this preference does not exist.
+     * @return Returns the preference value if it exists, or defaultValue. Throws
+     * ClassCastException if there is a preference with this name that is not
+     * a String.
+     *
+     * @throws ClassCastException If setting value's type is not String.
+     */
+    @Nullable
+    public String getString(@NonNull final String key, @Nullable final String defaultValue) {
+        awaitLoadFromDisk();
+        return map.containsKey(key) ? (String)map.get(key) : defaultValue;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **91**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **144**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).